optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraint.java

```diff
@@ -57,21 +61,81 @@ public DroolsConstraint(DroolsConstraintFactory<Solution_> constraintFactory,
-        for (DroolsFromUniConstraintStream<Solution_, Object> fromStream : fromStreamList) {
-            fromStream.createRuleItemBuilders(ruleItemBuilderList, scoreHolderGlobal);
+    private Set<DroolsAbstractConstraintStream<Solution_>> assembleAllStreams() {
+        final Set<DroolsAbstractConstraintStream<Solution_>> streamSet = new LinkedHashSet<>();
+        for (DroolsAbstractConstraintStream<Solution_> fromStream : fromStreamList) {
+            streamSet.addAll(assembleAllStreams(fromStream));
         }
-        return PatternDSL.rule(constraintPackage, constraintName)
-                .build(ruleItemBuilderList.toArray(new RuleItemBuilder<?>[0]));
+        return streamSet;
+    }
+
+    private Collection<DroolsAbstractConstraintStream<Solution_>> assembleAllStreams(
+            DroolsAbstractConstraintStream<Solution_> parent) {
+        final Set<DroolsAbstractConstraintStream<Solution_>> streamSet = new LinkedHashSet<>();
+        streamSet.add(parent);
+        for (DroolsAbstractConstraintStream<Solution_> child : parent.getChildStreams()) {
+            streamSet.addAll(assembleAllStreams(child));
+        }
+        return streamSet;
+    }
+
+    /**
+     * Return streams which have a given stream as a child.
+     *
+     * @param streams streams to process
+     * @param child stream to look for
+     * @return streams from the given collection whose {@link DroolsAbstractConstraintStream#getChildStreams()} contains
+     * the given stream
+     */
+    private Set<DroolsAbstractConstraintStream<Solution_>> predecessors(
+            Collection<DroolsAbstractConstraintStream<Solution_>> streams,
+            DroolsAbstractConstraintStream<Solution_> child) {
+        Set<DroolsAbstractConstraintStream<Solution_>> predecessorSet = new LinkedHashSet<>();
+        streams.stream()
+                .filter(stream -> stream.getChildStreams().contains(child))
+                .forEach(parent -> {
+                    predecessorSet.addAll(predecessors(streams, parent));
+                    predecessorSet.add(parent);
+                });
+        return predecessorSet;
+    }
+
+    /**
+     * Each constraint applies to a single scoring stream.
+     * From this stream, it needs to create rules for all of its parent streams.
+     * The method assembles the whole tree of streams, where the scoring node is the root and the from streams are the
+     * leaves.
+     * @param root scoring node for this constraint
+     * @return all streams which have the root as a child, transitively
+     */
+    private Set<DroolsAbstractConstraintStream<Solution_>> assembleScoringStreamTree(
+            DroolsAbstractConstraintStream<Solution_> root) {
+        Set<DroolsAbstractConstraintStream<Solution_>> allStreamSet = assembleAllStreams();
+        Set<DroolsAbstractConstraintStream<Solution_>> allPredecessorSet = predecessors(allStreamSet, root);
+        allPredecessorSet.add(root);
+        return allPredecessorSet;
+    }
```

Comment:
I had to add this method to the rule construction to fix the bug I discovered when I wrote tests for sharing streams.
(Both scoring streams would originally end up coming from the same constraint, because each constraint would originally be applied to each scoring stream, and due to the caching, the second constraint would never get a rule.)